include/spotify/json/codec/string.hpp

```diff
@@ -115,14 +115,36 @@ class string_t final {
     detail::fail(context, "\\u must be followed by 4 hex digits");
   }
 
-  static void decode_unicode_escape(decode_context &context, std::string &out) {
+  static void decode_unicode_escape(decode_context &context,
+                                    std::string &out,
+                                    unsigned high_surrogate) {
     detail::require_bytes<4>(context, "\\u must be followed by 4 hex digits");
     const auto a = decode_hex_nibble(context, *(context.position++));
     const auto b = decode_hex_nibble(context, *(context.position++));
     const auto c = decode_hex_nibble(context, *(context.position++));
     const auto d = decode_hex_nibble(context, *(context.position++));
     const auto p = unsigned((a << 12) | (b << 8) | (c << 4) | d);
-    encode_utf8(out, p);
+    if (json_unlikely(high_surrogate != 0)) {
```

Comment:
I think this is enough code to break out into a separate function. Also, for Visual Studio, it is better to order he if-statements in most likely order, it generates more efficient code. Clang and GCC uses the `json_[un]likely` macros.